crates/bevy_render/src/render_phase/mod.rs

```diff
@@ -1357,3 +1617,127 @@ impl BinnedRenderPhaseType {
         }
     }
 }
+
+impl RenderBin {
+    /// Creates a [`RenderBin`] containing a single entity.
+    fn from_entity(entity: MainEntity) -> RenderBin {
+        let mut entities = IndexSet::default();
+        entities.insert(entity);
+        RenderBin { entities }
+    }
+
+    /// Inserts an entity into the bin.
+    fn insert(&mut self, entity: MainEntity) {
+        self.entities.insert(entity);
+    }
+
+    /// Removes an entity from the bin.
+    fn remove(&mut self, entity_to_remove: MainEntity) {
+        self.entities.swap_remove(&entity_to_remove);
+    }
+
+    /// Returns true if the bin contains no entities.
+    fn is_empty(&self) -> bool {
+        self.entities.is_empty()
+    }
+
+    /// Returns the [`IndexSet`] containing all the entities in the bin.
+    #[inline]
+    pub fn entities(&self) -> &IndexSet<MainEntity, EntityHash> {
+        &self.entities
+    }
+}
+
+/// An iterator that efficiently finds the indices of all zero bits in a
+/// [`FixedBitSet`] and returns them in reverse order.
+///
+/// [`FixedBitSet`] doesn't natively offer this functionality, so we have to
+/// implement it ourselves.
+#[derive(Debug)]
+struct ReverseFixedBitSetZeroesIterator<'a> {
+    /// The bit set.
+    bitset: &'a FixedBitSet,
+    /// The next bit index we're going to scan when [`Iterator::next`] is
+    /// called.
+    bit_index: isize,
+}
+
+impl<'a> ReverseFixedBitSetZeroesIterator<'a> {
+    fn new(bitset: &'a FixedBitSet) -> ReverseFixedBitSetZeroesIterator<'a> {
+        ReverseFixedBitSetZeroesIterator {
+            bitset,
+            bit_index: (bitset.len() as isize) - 1,
+        }
+    }
+}
+
+impl<'a> Iterator for ReverseFixedBitSetZeroesIterator<'a> {
+    type Item = usize;
+
+    fn next(&mut self) -> Option<usize> {
+        while self.bit_index >= 0 {
+            // Unpack the bit index into block and bit.
+            let block_index = self.bit_index / (Block::BITS as isize);
+            let bit_pos = self.bit_index % (Block::BITS as isize);
+
+            // Grab the block. Mask off all bits above the one we're scanning
+            // from by setting them all to 1.
+            let mut block = self.bitset.as_slice()[block_index as usize];
+            if bit_pos + 1 < (Block::BITS as isize) {
+                block |= (!0) << (bit_pos + 1);
+            }
+
+            // Search for the next unset bit. Note that the `leading_ones`
+            // function counts from the MSB to the LSB, so we need to flip it to
+            // get the bit number.
+            let pos = (Block::BITS as isize) - (block.leading_ones() as isize) - 1;
```

Comment:
I don’t understand why leading_ones() is used here. Doesn’t the algorithm scan from the end of the set backwards to find zeros? Shouldn’t it look at trailing ones?

Reply:
Yeah, so `fixedbitset` actually fills each block from LSB to MSB, so bit 0 is `(1 << 0)`, bit 1 is `(1 << 1)`, etc. `leading_ones` searches from MSB to LSB. So we need to use `leading_ones` to iterate in reverse. (`trailing_ones` would search from LSB to MSB, which would be wrong in this case.)

Reply:
Ahhh, ok. I guess they did that for simpler indexing? Regardless, fine.